src/DX12/RHI/DX12CommandList.cpp

```diff
@@ -776,6 +774,29 @@ void DX12CommandList::Copy(RHITexture& src, RHIBuffer& dst, std::span<const Buff
     }
 }
 
+QueryHandle DX12CommandList::BeginTimestampQuery()
+{
+    QueryHandle handle = queryPool->AllocateQuery(GetType());
+    commandList->EndQuery(queryPool->GetNativeQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, handle.GetIndex() * 2);
+    queries.push_back(handle);
+    return handle;
+}
+
+void DX12CommandList::EndTimestampQuery(QueryHandle handle)
+{
+    commandList->EndQuery(queryPool->GetNativeQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, handle.GetIndex() * 2 + 1);
+}
+
+void DX12CommandList::ResolveTimestampQueries(u32 firstQuery, u32 queryCount)
+{
+    commandList->ResolveQueryData(queryPool->GetNativeQueryHeap(),
+                                  D3D12_QUERY_TYPE_TIMESTAMP,
+                                  firstQuery,
+                                  queryCount,
+                                  queryPool->GetTimestampBuffer().GetRawBuffer(),
+                                  firstQuery * sizeof(u64));
+}
```

Comment:
This should probably check (and assert) that the command list is closed, since otherwise there's no chance of being able to resolve.